invenio_rdm_records/services/pids/manager.py

```diff
@@ -87,15 +78,56 @@ def _validate_identifiers(self, pids, errors):
         for scheme, id_ in identifiers:
             pids[scheme]["identifier"] = id_
 
+    def _validate_pids(self, pids, record, errors):
+        """Validate an iterator of PIDs.
+
+        This function assumes all pid schemes are supported by the system.
+        """
+        for scheme, pid in pids.items():
+            provider = self._get_provider(scheme, pid.get("provider"))
+            success, val_errors = provider.validate(record=record, **pid)
+            if not success:
+                errors.append({"field": f"pids.{scheme}", "messages": val_errors})
+
+    def _validate_record(self, pids, record, errors):
+        """Validate the record according to the PIDs' requirements.
+
+        Here we check if the record is compatible from the point of
+        view of the pids...
+            - ... it contains
+            - ... it would contain according to configured required pids
+
+        The responsibility lies with each provider since they are the ones
+        that know their criteria for a record that is complete enough to get
+        a PID.
+        """
+        schemes = set(pids.keys()) | set(self._required_schemes)
```

Comment:
Solves a subtle potential issue compared to previous PR where 2 different providers could have been listed for the same scheme (the default and the selected) or same provider listed twice.

Reply:
I guess that if a required scheme is not in `pids`, we should fail?
Why not validating this before, very explicitly, and probably in `_validate_pids`?

If this is a subtle use case, it should be made very clear and very explicit, e.g. documenting it in the code or via a named var/func.
The code with the extra comments `# provider_name for an absent-but-required pid will be None` basically hides all this.

Reply:
> 

No: if a required scheme is not in pids it should not necessarily fail. We may be in the case where a new draft is made/published and doesn't have the required PID yet (e.g. doesn't have `"doi": {...}` in its "pids" fields yet). The test for this case already exists here:
https://github.com/inveniosoftware/invenio-rdm-records/blob/87b986241e53f8741b4ab9a457b98b26bfb6e2a7/tests/services/components/test_pids_component.py#L351 .

For the `provider_name` being `None`, that's because in that case the `"provider"` value is not given. By passing `None` we get the default provider for the scheme to be chosen.

Reply:
>  if a required scheme is not in pids it should not necessarily fail... a new draft is published and doesn't have the required PID yet...
If published, it must have it, no?

In any case, honestly, the fact that we need to align on the implemented behaviour with PR comments is a hint that we should improve existing or new implementations, independently who coded it, to make it more readable/understandable/immediate: no one wants to spend hours trying to figure out what a few lines of code are solving.